lib/apiservers/engine/backends/commit.go

```diff
@@ -0,0 +1,472 @@
+	"github.com/docker/docker/api/types/backend"
+	containertypes "github.com/docker/docker/api/types/container"
+	eventtypes "github.com/docker/docker/api/types/events"
+	"github.com/docker/docker/builder/dockerfile"
+	dockerimage "github.com/docker/docker/image"
+	dockerLayer "github.com/docker/docker/layer"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/streamformatter"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/docker/reference"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/cache"
+	"github.com/vmware/vic/lib/apiservers/portlayer/models"
+	"github.com/vmware/vic/lib/imagec"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/version"
+	"github.com/vmware/vic/pkg/vsphere/sys"
+)
+
+// Commit creates a new filesystem image from the current state of a container.
+// The image can optionally be tagged into a repository.
+func (i *Image) Commit(name string, config *backend.ContainerCommitConfig) (imageID string, err error) {
+	defer trace.End(trace.Begin(name))
+
+	// Look up the container name in the metadata cache to get long ID
+	vc := cache.ContainerCache().GetContainer(name)
+	if vc == nil {
+		return "", NotFoundError(name)
+	}
+
+	// get container info
+	c, err := containerEngine.ContainerInspect(name, false, "")
+	if err != nil {
+		return "", InternalServerError(err.Error())
+	}
+	container, ok := c.(*types.ContainerJSON)
+	if !ok {
+		return "", InternalServerError(fmt.Sprintf("Container type assertion failed"))
+	}
+	if container.State.Running || container.State.Restarting {
+		return "", ConflictError(fmt.Sprintf("%s does not support commit of a running container", ProductName()))
+	}
+	// TODO: pause container after container.Pause is implemented
+	newConfig, err := dockerfile.BuildFromConfig(config.Config, config.Changes)
+	if err != nil {
+		return "", err
+	}
+
+	if config.MergeConfigs {
+		if err := merge(newConfig, container.Config); err != nil {
+			return "", err
+		}
+	}
+
+	rc, err := containerEngine.containerProxy.GetContainerChanges(vc, true)
+	if err != nil {
+		return "", fmt.Errorf("Unable to initialize export stream reader for container %s", name)
+	}
+
+	ic, err := getImagec(config)
+	if err != nil {
+		return "", err
+	}
+
+	layer, err := downloadDiff(rc, container.ID, ic.Options)
+	if err != nil {
+		return "", fmt.Errorf("Unable to export stream reader for container %s: %s", name, err)
+	}
+	if err = setLayerConfig(layer, container, config, newConfig); err != nil {
+		return "", err
+	}
+	// Dump metadata next to diff file
+	destination := path.Join(imagec.DestinationDirectory(ic.Options), layer.ID)
+	err = ioutil.WriteFile(path.Join(destination, layer.ID+".json"), []byte(layer.Meta), 0644)
+	if err != nil {
+		return "", err
+	}
+	imagec.LayerCache().Add(layer)
+
+	var layers []*imagec.ImageWithMeta
+
+	layers = append(layers, layer)
+	lm := layer
+	for pl := lm.Parent; pl != imagec.ScratchLayerID; pl = lm.Parent {
```

Comment:
This for loop is fine, but it was strange to look at initially.  I'd suggest the following.  It's up to you whether you use it.

```
	pl := lm.Parent
	for {
 		if pl == imageC.ScratchLayerID {
			break
 		}

 		// populate manifest layer with existing cached data
 		if lm, err = imagec.LayerCache().Get(pl); err != nil {
 			return "", InternalServerError(fmt.Sprintf("Failed to get parent image layer %s: %s", pl, err))
 		}
 		layers = append(layers, lm)

 		pl = lm.Parent
 	}
```